### ISSUE TITLE
Swappable themes using global

### ISSUE THREAD

**Issue by avanderhoorn** [css-modules/css-modules#160]
Current theming solutions seem to either rely on you passing around a theme object or using css variables. The former is rather annoying to do and the latter doesn't have cross browser support. 

When thinking about the theme object approach, besides being annoying, it requires that your component know not only about the local style (i.e. `styles.btn`) but also about the theme style (i.e. `theme.btn`) resulting in code that looks like `<div className={classNames(styles.btn, theme.btn)}>`. To me, the fact that the component has this knowledge feels like a bit of a code smell and like the knowledge is in the wrong place.

What i would prefer to do is something like the following (note its not perfect but a work in progress):
```
//index.css 
.btn {
    font-weight: bold;
    composes: normalTextColor from "./theme.default.css";
}

//theme.default.css
:global {
    .normalTextColor { } // purposefully left blank as no default is supplied
}

//theme.dark.css
:global {
    .normalTextColor { color: green; } 
}

//theme.light.css
:global {
    .normalTextColor { color: red; } 
}

//index.js
<div className={styles.btn}>

//generated html from index.js
<div class="index__btn___113Qb normalTextColor">
```
To make this work as expected (when either the dark or light theme is applied) you load/unload either `theme.dark.css` or `theme.light.css`, or load in both themes and just have the selectors scoped to `.dark` or `.light` which is removed from an upstream class.

By doing what I have done, the knowledge of what theme classes should come together to make the logical `btn` are encapsulated in the components css file. Now, I will be the first to admit that its not a perfect solution, but I do think its better than passing a theme object around and encapsulates knowledge. Ultimately, I think css variables are the better solution but I can't use them until edge starts supporting them... hence looking for other solutions.

Besides discussing the approach, unfortunately I haven't been able to get the above to work as the system seems to be ignoring my `:global`s and generates a unique name for each usage of `normalTextColor`. Any thoughts or ideas?

//@geelen - as per twitter conversation

**Comment by avanderhoorn** [comment]
Another idea that I had which is more involved then just trying to get the above working for the here and now, and would be a great help for those that couldn't use CSS Variables.

This follows up on the concept of how we can take advantage of CSS Variables now to help automatically produce the theme files. Typically when you talk about using CSS Variables and pollyfills, you end up with a compiled version of you css files which just have the variables replaced. From a theming perspective this doesn't really work and in a naive implementation, you would end up with two css files (one for each theme) which all the classes are repeated and just the colors that are different. But if we do some code generation, I wonder if we can make things smarter.

Lets look at the sample I did above (slightly modified). Its probably the best we can do without being able to use CSS Variables at runtime.
```
//index.css 
.btn { font-weight: bold; }

//theme.dark.css
.btn { color:green; } 

//theme.light.css
.btn { color: red; } 
```
Now putting aside that this doesn't really work from a CSS Modules perspective (as mentioned in the original issues), this approach works but ultimately end up being brittle and hard to maintain.

Taking the fact that the above is what we ultimately would like to have, how can we make the authoring experence easier, less brittle and easier to maintain. What if we could author something like the following:
```
//index.css 
.btn { 
    font-weight: bold; 
    color: --main-color;
}

//theme.dark.css
--main-color: green;

//theme.light.css
--main-color: red;
```
Now you might be thinking that this looks just like normal usage of CSS variables, but remember we can't use CSS variables at runtime yet due to browser support and standard CSS variables end up just bloating things.

But what if we had a smart post processor which scanned the target css files, and when it saw the use of a css variable, automatically generated the ideal css from the above that we want. 

To do this, when the processor detects the usage of a variable it would: 
 - strip styles that use vas (in this case `color: --main-color;`), 
 - look to see what selectors are associated with the style (in this case `btn`),
 - take the registered themes and lookup what the variable is (in this case `green` and `red`), and generate a new class definition

Hence at the end of the process we would have the below, but have not had to manage the same definitions being duplicated in multiple places and the usage/definitions are centralized in a way that makes sense. 
```
//index.css 
.btn { font-weight: bold; }

//theme.dark.css
.btn { color:green; } 

//theme.light.css
.btn { color: red; } 
```

As a side point, I would probably suggest that the selectors be scoped to the theme they are targeting. This makes runtime switching easier as you only have to replace the class on a root node in the app. Hence it would actually look like the following:
```
//index.css 
.btn { font-weight: bold; }

//theme.dark.css
.dark .btn { color:green; } 

//theme.light.css
.light .btn { color: red; } 
```

Lastly, from a CSS Modules perspective, because css "names" are generated, it would obviously need to use the generated names, etc, etc.

Here is a more complex scenario:
```
//index.css 
.btn { 
    font-weight: bold; 
    color: --main-color;
    border: 1px solid --main-color;
}
.btn-inner {
    margin: 10px;
    padding: 10px 20px
    background-color:--main-light-color;
}
//theme.dark.css
--main-color: green;
--main-light-color: light-green;

//theme.light.css
--main-color: red;
--main-light-color: light-red;
```
and here is the generated output:

```
//index.css 
.btn { 
    font-weight: bold; 
}
.btn-inner {
    margin: 10px;
    padding: 10px 20px
}
//theme.dark.css
.dark .btn { 
    color: green;
    border: 1px solid green;
}
.dark .btn-inner {
    background-color: light-green;
}

//theme.light.css
.dark .btn { 
    color: red;
    border: 1px solid red;
}
.dark .btn-inner {
    background-color: light-red;
}
```

This is approach is also nice because the way in which people are authoring their CSS is forward compatible as more browser support comes online.

**Comment by avanderhoorn** [comment]
As a side note, I've created a prototype plugin based on the above logic and it actually works really well. If you are interested I can share with a sample.

**Comment by geelen** [comment]
Thanks for your suggestions! While I'm a little wary of co-opting the CSS Variables syntax in this way, I can see how it does make sense, and something about converting variable declarations into extra CSS (adding `.dark` or `.light` in front of rules) might work really well. Can you share your sample?

The other thing I was thinking is whether the [Web Components Polyfill](https://github.com/WebComponents/webcomponentsjs) could be employed here. I know that the Polymer project [uses custom properties for styling](https://www.polymer-project.org/1.0/start/first-element/step-5) but I'm not sure how. I might jump on Edge when I'm on my home PC tonight and have a look.

**Comment by avanderhoorn** [comment]
Thanks for getting back to me. Here is a quick gist of what I'm doing atm - https://gist.github.com/avanderhoorn/1fed982b1bfe5ac016cc2fdcc9d2faeb. I can get a sample up and running in the morning but this should give a general idea.

**Comment by avanderhoorn** [comment]
Note, the current implementation does assume that you are using `ExtractTextPlugin` as it taps the pipe post that asset being created and hence the local css selectors are already changed to their unique counter parts. I'm sure this can be done better but it shows the point.

**Comment by dvkndn** [comment]
Hi @avanderhoorn and @geelen , I hope this is still in discussion. I recently think of another solution for theming while using CSS Modules for our production web app:

Our solution is that we use **lower level** classes for styles that changed between themes, using a (optional) global class to seperate them, and our real components compose these lower level classes.

```
/* color.css */
.text {
  color: black;
}
:global(.dark) .text {
  color: white;
}
```

```
/* button.css */
.main {
  composes: text from "./color.css";
}
```

```
/* button.jsx */
import styles from '../styles/button.css';
...
<button className={styles.main} />
...
```

Now all buttons' text will be black by default. However, if the `body` tag have a 'dark' class, then all buttons' text will be white.

**Comment by EvNaverniouk** [comment]
@dvkndn Unfortunately your solution has a major problem. Having two levels of classes makes it much harder to have components override default theme styles. Consider this scenario:

The "dark" theme defines this style for all buttons:

```
.dark .button {
    background: black;
}
```

Now in your button component you want to change the background.

```
.button {
    background: #400;
}
```

Unfortunately, that style override won't apply because its specificity is lower than what is defined in the theme file. So you have to either know which theme is set and do `.black .button {}` or you have to increase specificity some other way, ie `.button.button {}`. Both of those are very awkward and prone to developer confusion.

For this reason, I think adding specificity for a theme via a class name on the `<body />` tag is not ideal.

**Comment by geelen** [comment]
Yeah, specificity is going to bite you with this approach, since you're using the cascade to do something that really should be inherited. For example,

```css
/* color.css */
.text {
  color: var(--text-color, black);
}
:global(.dark) {
  --text-color: white;
}
```

Then, if you had some component that wanted to `compose` your `text` class but also override the text colour (silly in this trivial case but could happen with more complex classes) you can do:

```css
/* button.css */
.main {
  composes: text from "./color.css";
  --text-color: red;
}
```

Since `--text-color` isn't set on the element, it gets inherited from `body`. If `body` has the `.dark` tag, it gets set to white, otherwise it's undefined and the default is black. But if you inject a value for `--text-color` on the `.main` element _or anywhere above it_ it gets applied.

We're at [64% for CSS Variables support](http://caniuse.com/#feat=css-variables) which is so so close to being usable, but not quite yet :(

**Comment by dvkndn** [comment]
Thank you two. I haven't thought about it because we didn't need to override the colour..

We actually provide **all** possible colours in `color.css` (a light one and a dark one for each), then only compose at the **final** components (which means, we don't compose at general or middle components like "button", but something at the end, most special, like "comment-button".

Well, I guess it's not a ideal solution because we need to define the color whenever we create a new type of button. There is no default color for all buttons (actually, it is inherited).

```css
/* color.css */
.text {
  color: black;
}
:global(.dark) .text {
  color: white;
}
.positive {
  color: green;
}
:global(.dark) .positive {
  color: darkgreen;
}
/* button.css */
.main {
  /* don't compose color here */
}
/* comment-button.css */
/* lowest level, no thing inherit from this */
.main {
  composes: positive from "./color.css";
}
```

**Comment by TheSisb** [comment]
@EvNaverniouk Unless I'm misunderstanding, I don't think specificity is an issue with @dvkndn's solution.  From what I gather s/he composes the `color` or `background-color` he wants.  Therefore there is no need to overwrite it - just omit the `composes` rule.

**Comment by dvkndn** [comment]
@TheSisb yes thank you for pointing out. To be more clear, my idea is that we will have different classes for different colors (eg:`.text`, `.secondary`, `.accent`...), then the component **chooses** the color it wants to use by composes the corresponding class. Global classes (eg: `.dark`) help define what is the actual value of that color/class in each situation. I don't need to override anything so far.

**Comment by dvkndn** [comment]
@TheSisb I agree that the most advantage I have here is that it is seperate the components from the theming system. However, I don't know why you want to omit the `compose`? Because I think it would be easier if we defined style by classes, then components compose those classes and themes defines those same classes?

Like this:
**Step 1:** define classes for colors and how it will look like in different theme
```css
color.css
.text {
  color: black;
}
:global(.dark) .text {
  color: white;
}
```
**Step 2:** compose class in component when we want to style the color
```css
sidebar.css
.main {
  composes: text from "color.css";
}
```

**Comment by TheSisb** [comment]
In my project I use colors.css to define variables like 

```
$brand_blue: blue;
$brand_red: red;
```

Then in my components when I need a color I just say what color I want:
```
.sidebar {
  background: $brand_blue;
}
```

Then when my sidebar is themed, I would do:

```
:global(.dark) .sidebar {
  background: $brand_red;
}
```

I prefer this approach because:
- I don't have to keep looking up what color is actually used for ".sidebarBackground" by jumping down an extra file.
- Removes the need for an extra class in my HTML markup and in my CSS codebase.  Color variables get replaced at compile time.
- For more complex projects and I'm sure you've encountered this, if I make a text color maybe 90% of my app will use it.  For the next 10% I need to make 5+ variants for one-time use components.  For example

```
// :(
.main {
  composes: textColor, backgroundColor from 'colors.css';
}
.secretBar {
  composes: secretBarTextColor, secretBarBackgroundColor from 'colors.css';
}
.otherThings {
  // etc...
}

// :)
.main {
  color: white;
  background: black;
}

.secretBar {
  color: $grey-1;
  background: $grey-7
}
```

- I like to keep all the CSS belonging to my components in the component's CSS file.  In my case I define the theme in the current file.  In your case, your dependency files (ie: colors.css) have the theme rewrites. 
- I don't think this was how `composes` was intended to be used. I want to use `composes` strictly for useful compositions rather than overwrites.

**Comment by dvkndn** [comment]
Thank you, now I understand why you don't use compose. Yes it is a matter of preference and maybe our situation too. It sure turn out that our approach is quite different.

Two biggest difference I think is that first, your themes are in new, separate files, while all my themes is in 1 single file, and second, I use composes almost everywhere. I think I'm all in for the idea of using classes to build styles when I saw `composes`. My production HTML is full of element with hashed classes (like `.zh18s .j38hf .s8asn...`) and each class only do 1 thing.

**Comment by pleunv** [comment]
I'm very interested in an approach like the second one proposed by @avanderhoorn. Moves away theming responsibility from the components themselves while keeping overhead small, and still offers a lot of flexibility.

**Comment by fahad19** [comment]
Could be related: 

* Issue: https://github.com/css-modules/css-modules/issues/227
* PR: https://github.com/webpack-contrib/css-loader/pull/496

**Comment by sompylasar** [comment]
https://github.com/react-theming/react-theme-provider
